src/wayland/xwayland_keyboard_grab.rs

```diff
@@ -0,0 +1,176 @@
+use wayland_protocols::xwayland::keyboard_grab::zv1::server::{
+    zwp_xwayland_keyboard_grab_manager_v1::{self, ZwpXwaylandKeyboardGrabManagerV1},
+    zwp_xwayland_keyboard_grab_v1::{self, ZwpXwaylandKeyboardGrabV1},
+};
+use wayland_server::{
+    backend::{ClientId, GlobalId, ObjectId},
+    protocol::{wl_seat, wl_surface},
+    Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New, Resource,
+};
+
+use crate::{
+    backend::input::KeyState,
+    input::{
+        keyboard::{self, KeyboardGrab, KeyboardInnerHandle},
+        SeatHandler,
+    },
+    utils::Serial,
+    xwayland::XWaylandClientData,
+};
+
+const MANAGER_VERSION: u32 = 1;
+
+pub trait XWaylandKeyboardGrabHandler {
+    fn grab(&mut self, surface: wl_surface::WlSurface, seat: wl_seat::WlSeat, grab: XWaylandKeyboardGrab);
+}
+
+// TODO
+pub struct XWaylandKeyboardGrab {
+    grab: ZwpXwaylandKeyboardGrabV1,
+    surface: wl_surface::WlSurface,
+}
+
+impl<D: SeatHandler + 'static> KeyboardGrab<D> for XWaylandKeyboardGrab {
+    fn input(
+        &mut self,
+        data: &mut D,
+        handle: &mut KeyboardInnerHandle<'_, D>,
+        keycode: u32,
+        state: KeyState,
+        modifiers: Option<keyboard::ModifiersState>,
+        serial: Serial,
+        time: u32,
+    ) {
+        // TODO set_focus?
```

Comment:
Yeah I am not sure, if the KeyboardHandler does this when setting the grab already, but we definitely want to set focus to `None` at the start of the grab.

Reply:
Though this would send a wl_keyboard.leave event, I am not sure, if that is what Xwayland expects? Or if it wants the surface to get focus?

Reply:
From the protocol "forcing the keyboard focus for the given seat upon the given surface.",
I guess we want to call `set_focus` upon `self.surface` here.

This means we might need a requirement for `From<WlSurface>` for `<D as SeatHandler>::KeyboardFocus`. I wonder if that is too generic and we want to introduce a wrapper-type like `XwaylandGrabSurface(pub WlSurface)` here and specifically require a `From`-implementation for that type. This would allow compositors to special case this particular Focus-variant.

Reply:
If we define the helper trait as `trait XWaylandKeyboardGrabHandler: SeatHandler`, we can require a method returning `Self::KeyboardFocus`, which might be cleaner than using `From` with a special wrapper type if it's only used here.

I guess for either Anvil or Cosmic this `WlSurface` needs to first be converted to the corresponding `X11Surface`.

Reply:
Right the X11Surface would be cleaner to handle input properly.

I have no clear vision of how you would want to model this, but feel free to go ahead with this approach. It sounds sensible and I would like to see how it would look in practice.